bonfire/bonfire.py

```diff
@@ -1016,6 +1073,198 @@ def _cmd_apps_what_depends_on(
     print("\n".join(found) or f"no apps depending on {component} found")
 
 
+@reservation.command("create")
+@click.option(
+    '--bot',
+    '-b',
+    is_flag=True,
+    help="Use this flag to skip the duplicate reservation check (for automation)",
+)
+@options(_reservation_process_options)
+@options(_timeout_option)
+def _create_new_reservation(bot, name, requester, duration, timeout):
+    def _err_handler(err):
+        msg = f"reservation failed: {str(err)}"
+        _error(msg)
+
+    try:
+        res = get_reservation(name)
+        # Name should be unique on reservation creation.
+        if res:
+            raise FatalError(
+                f"Reservation with name {name} already exists"
+            )
+
+        res_config = process_reservation(name, requester, duration)
+
+        log.debug("processed reservation:\n%s", res_config)
+
+        if not bot:
+            if check_for_existing_reservation(res_config["items"][0]["spec"]["requester"]):
+                _warn_of_existing(res_config["items"][0]["spec"]["requester"])
+
+        try:
+            res_name = res_config["items"][0]["metadata"]["name"]
+        except (KeyError, IndexError):
+            raise Exception(
+                "error parsing name of Reservation from processed template, "
+                "check Reservation template"
+            )
+
+        apply_config(None, list_resource=res_config)
+
+        ns_name = wait_on_reservation(res_name, timeout)
+    except KeyboardInterrupt as err:
+        log.error("aborted by keyboard interrupt!")
+        _err_handler(err)
+    except TimedOutError as err:
+        log.error("hit timeout error: %s", err)
+        _err_handler(err)
+    except FatalError as err:
+        log.error("hit fatal error: %s", err)
+        _err_handler(err)
+    except Exception as err:
+        log.exception("hit unexpected error!")
+        _err_handler(err)
+    else:
+        log.info(
+            "namespace '%s' is reserved by '%s' for '%s'",
+            ns_name,
+            res_config["items"][0]["spec"]["requester"],
+            duration,
+        )
+        click.echo(ns_name)
+
+
+@reservation.command("extend")
+@click.option(
+    '--namespace',
+    '-ns',
+    type=str,
+    default=None,
+    help='Name of the reserved namespace',
+)
+@options(_reservation_process_options)
+def _extend_reservation(name, namespace, requester, duration):
+    def _err_handler(err):
+        msg = f"reservation extension failed: {str(err)}"
+        _error(msg)
+
+    try:
+        res = get_reservation(name, namespace)
+        if res:
+            res_config = process_reservation(name, res["spec"]["requester"], duration)
```

Comment:
I tried to extend for an ns that does not exist and the error didn't seem to get handled cleanly:

```
❯ bonfire reservation extend -ns foo
2021-09-23 15:26:10 [    INFO] [          MainThread] Retrieving reservation
2021-09-23 15:26:11 [   ERROR] [          MainThread] hit unexpected error!
Traceback (most recent call last):
  File "/home/bsquizza/repos/bonfire/bonfire/bonfire.py", line 1156, in _extend_reservation
    res_config = process_reservation(name, res["spec"]["requester"], duration)
KeyError: 'spec'
ERROR: reservation extension failed: 'spec'
```

Not sure how we got to line 1156, would have thought `if res` would have evaluated as `False`